sdk/core/core-util/src/jsonMergePatch.ts

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { isObject } from "./object";
+
+export type JsonMergePatch<T> = T extends { [k: string]: unknown }
```

Comment:
I'm trying this out in the [playground](https://www.typescriptlang.org/play?target=6#code/PQKhCgAIUhBSDmBTAdkgTgSwMaQM4AWAhgA5KQBmA9upEZCQDZGYqQBSAypFQEYBWSbABcAdFBDBwSAB4kawyMICeZSAFUUAaxRUA7igDyAoYoC8kAN6QA2ngBc+YVhQIAuo4Cu23QcgBfAG5wcFAIaEgACSRGMlphKkgAEyRhDABbVnI9AlQ6NlYST0VMPDpEVAwcDm4+QRFxCIABdFTPdBQy509s3IKUIpKyojY60yVVcmFiIchdRRRPRkYAGjh0dCJlNYAlJAQAUTk1mkgAESI0xslpOQVKbxFMKgK8Y3rhAApC4q8ffRQAEpHD9ZpodAD3uNLFBIK1hO02J9YZAJmQqBRIKDIGZcZAAERjET4yAAMlJKKxA2KkAAhHjFssyRTUajabANltRKUOZtlN9qcJAczKbSBYMqXhhCNsEgMZA9oc5MLyaLxTTWFKZXLMRc0oDYYDgv4QrJ5OhFCo1Ow8C8ALIYZAABUu2AIAB4ACoAPhxkE9kFkaRQSTK1hsWkcUpc7j+EL8JtRAH4OLaUA70M7XQQoSIvd7YY5gMBIABJTEB0ooADkihGPBMIjWenI-E8UsgykwMSS-spnuC4Ct5Bt9sdSBdwjdueEXsDMmDoastkjThjHkg3njbH8vrMsOL-tycKE7TwzzYskwwjKOTy03IRMUSSoSDwNcU2Be0tYa9YCAAWi0JBlCQXsSHQKg4mEbs8FhGFWUgQ8AGFcmwLQykwTF6AgqCMBULEyigmCXiIRgeFobwUgoLJezI1oiCSZRIE+XRID0LYlESJJShg1xPFKAglFydI4AAMVgUsAGkDUQmwpKpSBgOUeVPTcJNHFHdNx0nN0nUg6DlC9eS3G9YJUSCEJh1TMdMwnbN9Lwi0jJ9P0qKQGi0F7INUCXT1YRTQ89mwM83zWRieNcOYlnIhJhKmSZNxQC8eExF5GGYh8eBIEiUDIhgDPw2DKS0jMsynD0ADkXkqmKiF4RgkHzX0AB9ouWQskJLct2vI0o6EYBimIKt9UGENZpn6iDRvQAA3N8sWEEq0zK+yKvdaqUFq5Z6sa5rIDaz4A1JXqjRCMIJDgEgmFgyAAAMSGzO6uPu6U7OEZ6iGwL90EihAXqyrhDEqyBVsgXShJ2MTkOrMo8DIbBRHOV8ynmE90ioebLtaABHATWl7Giezwa5ICaR7NlEt7kEUACjymIh3tS+KGGzUnycZohRMeirIDpz1jx5t0kYF8hPW4ay9EwJlpvmlBLWPVoMdm-L5Vx-GwMobtGCXChINEp9LuHPA1l4Gk2LwTw3UgFXGEwJJLhS-qXzQRSsooR5csgXgqCY9n4URMo6Z6rKhaEkZe2p1I6FaOYXgAxm+QbD4TZPBEOi6QXs01qOxEgQwH3QKW8CQbG2gz+6w7u64pDNe4PZQJ4XkgdIdOzfNPlzxxPTWMPNJWtv1p9YF-SsWEA46FvB7dThnBwL5c977M1k+EgR5IQjR7MX1uiQM6TVCMBLtga67YWh6npeu7c8+76aD+gHjyBkGwYhhVodh-AEaRs4UbjxQlaY1LhEdWmACZa2JuzCmXMlCMxpnzemsCmbylDmzS6HNKas15vzLOFURbHnFmibI0tyKyzGizQBttmagPAUTHWZQ9ZUANo2JaERjam3NokS21tbb20ds3Z2LxyC-ndp7FKPs-boOgVTSYABxTwjNex03gNgMijBeBfS0C9XCs17bkGoPEY8xAQx2yivKVRJdby5Fjkga81jXpwNSM9Ggl0L4VWev1NG9YnxIxDgQfqqjlgaPQpvNGOi9FJAmseBuTc2AT06PddynkwJ3UuneNgIxmLYirLWcougUAJ05MxHxR5+oxK9oQKgSxezxPKN0aYzFbY9B4CgDK7E+iuKrpvegltry7X0Wo4JWiDEszDjnRxedOAIywjgNR2xhJlLEc3YgqNEgcWUEbRIX0vzeEUCMzIeBzxRRoZrOhS5Lhwl2ZgVuaxbQLIYUstglTqlp0RJQMiJd2moEutwoSZzhixwOUchA-ty4JODpiVBvMI5IPgYzcg+TClJyfKneJmdyBjMjhM0mfiEpqAQAo36m9qK0XyDUsFwwlDoGKAQRpZFmkjLcW6O6aw0WV0vkQPAAFSikz2OnBJSTSVUELsXJANdbjmj2Y8qedkIazywHmH0yJURd39CsWEfcbLaVle3H06qVVyMJUkRwq9szdwOlq1aEN8zCm3lg62-V-IjwHBK+u0rW46oqvK+eHdVU93tQQfutlypuhteatqgqvLBDrhaB4jcvYepDQQb1irvTKthakbu+qA1Bu1Um-M2bhzyMURpFimqAxtVKtPD0w8cS+jDpvJ14akokq8mPVEWEWK0lKDOTuEzhQtTat2t4LDTUVUBMKBCrJalFqNUmUQY63TChTA2v4rawLmQCCEVEX5OiKGUngT0VA9jK3IBYGcogxpYDfIuggslWSiBoowNI6BPifBsAAfTWE0pAbhbW+h-TiBkMV72olEOkUgb6IwgTWB+v9dalIgTOrCXdHZMUHHlteso56WGXsw7BW9gJH3SxfVB5SsH4N2tpAeo9J6gHckbowTwKQ8CfGUhOwcqIRmfFQ4oK9xTMTofw2+SdlJeMrhgzbelv6-T8c3Tul4HY3TSySJ6CZfpc7QeUG4eTWJMTfBHR8Hj-idZqfehO9tiFAyYeUDYAAjG4P0ia1ozzniIYzKmzM02-dJugZQq2etDcOcxJnVMTO9GdRCiYt3jzBfnXDjD0gYbnjemwlJRAZYvVegjuciNPtI++8jkA4P-rpDR49SBT0MewExljbGkOAmzWBjLQmUsm1hPBzl-pjTnUPGkKUIRWAvooF9cgToMBpks6iPKrcoxz1cLp5gUpKpcyQHNmMunKssEYOt-8unhBS2EC+yIEdGqlujHtykCA7GeF4CdkxSBzvzYQLpogHFfoOCm1Z5wmjYAhjEtrJIpaHQO0YLpxCtpvoYGB2BMiumTQH2siD-KFgCQICoDrEkbV8TnkYPNdAWOCS8EgigAAXkgfEIRxOeBIA7NI430BplzVa7MM53QM7THuL7B3rzHdO2t3q4AD5AA) and ` T extends { [k: string]: unknown }` is not compatible with an interface such as 


```ts
interface Person {
    name: string;
    lastName: string;
    email: string;
    twitterHandle?: string;
    githubHandle?: string;
    awards: {
        trackAndField?: Medal;
        soccer?: Medal;
    }
}

type Medal =  "gold" | "silver" | "bronze"
```

however `T extends object` does work fine. Am I missing anything?

Reply:
I forgot that TS can be capricious when it comes to index signature types. The issue here is that arrays, tuples, Dates, functions, etc. all extend `object`, but their string properties shouldn't be in the patch type. 

Typically, this kind of type wizardry involves testing as many edge cases as we can find. I've just pushed a fix for this, but it now involves taking a runtime dependency :(